File: flutter/sceneview_flutter/ios/Classes/ARTapToPlace.swift
import Foundation
import simd

// MARK: - AR tap-to-place (#3780): pure logic
//
// Kept free of ARKit, RealityKit and Flutter so the math can be checked off-device.
// The Android bridge holds the same contract in `ARTapToPlace.kt`, and its JVM test
// (`ARTapToPlaceTest`) pins the wire format that both sides share with Dart.

/// How far a pinch can resize a placed model, relative to its placed size. The
/// same range as Android's `PLACED_SCALE_RANGE`, and the one the Dart
/// `placeModel` doc quotes.
let flutterPlacedScaleRange: ClosedRange<Float> = 0.25...4

/// Orientation of a plane hit, in the ARCore `HitResult.hitPose` convention
/// that the Dart `ARHitResult` documents for both platforms:
/// - +Y is the plane normal;
/// - +Z lies in the plane and points toward the camera;
/// - +X is `Y × Z`.
///
/// ARKit's raycast result has no such convention (its orientation follows the
/// plane anchor), so the iOS bridge derives it here to keep the Dart API
/// identical across platforms. A model placed with this rotation faces the user.
///
/// When the camera is straight above the hit, the direction to the camera has
/// no in-plane component. Any heading is then valid, so a fixed world axis
/// projected onto the plane is used instead of a NaN.
func flutterPlaneHitRotation(
    normal: SIMD3<Float>,
    hitPosition: SIMD3<Float>,
    cameraPosition: SIMD3<Float>
) -> simd_quatf {
    let normalLength = simd_length(normal)
    let y = normalLength > 1e-6 ? normal / normalLength : SIMD3<Float>(0, 1, 0)

    func inPlane(_ v: SIMD3<Float>) -> SIMD3<Float> { v - simd_dot(v, y) * y }

    var z = inPlane(cameraPosition - hitPosition)
    if simd_length(z) < 1e-4 {
        let reference: SIMD3<Float> = abs(y.z) < 0.9 ? SIMD3(0, 0, 1) : SIMD3(1, 0, 0)
        z = inPlane(reference)
    }
    z = simd_normalize(z)
    let x = simd_cross(y, z)
    return simd_normalize(simd_quatf(simd_float3x3(columns: (x, y, z))))
}

/// World transform of a hit pose: `rotation`, then a translation to `position`.
func flutterHitTransform(position: SIMD3<Float>, rotation: simd_quatf) -> simd_float4x4 {
    var transform = simd_float4x4(rotation)
    transform.columns.3 = SIMD4<Float>(position, 1)
    return transform
}

/// Encodes a plane hit as the `onPlaneTap` payload that the Dart
/// `ARHitResult.fromMap` decodes. Uses the same keys as Android's `planeHitMap`.
func flutterPlaneHitMap(
    id: String,
    position: SIMD3<Float>,
    rotation: simd_quatf,
    planeType: String,
    distance: Float
) -> [String: Any] {
    [
        "id": id,
        "x": Double(position.x),
        "y": Double(position.y),
        "z": Double(position.z),
        "qx": Double(rotation.imag.x),
        "qy": Double(rotation.imag.y),
        "qz": Double(rotation.imag.z),
        "qw": Double(rotation.real),
        "planeType": planeType,
        "distance": Double(distance),
    ]
}

/// A decoded `placeModel` call. It mirrors Android's `PlaceModelRequest`.
struct FlutterPlaceRequest {
    let hitId: String?
    let position: SIMD3<Float>
    /// Unit quaternion. A zero or non-finite input decodes to identity.
    let rotation: simd_quatf
    let modelPath: String
    /// Size of the model's largest dimension, in metres (`ModelNode.scale`).
    let size: Float
    let editable: Bool
    let draggable: Bool
    let rotatable: Bool
    let scalable: Bool

    var canDrag: Bool { editable && draggable }
    var canRotate: Bool { editable && rotatable }
    var canScale: Bool { editable && scalable }

    /// Decodes `{hit: ARHitResult.toMap(), model: ModelNode.toMap(), editable,
    /// draggable, rotatable, scalable}`. Returns nil when the hit or the model
    /// path is missing.
    init?(arguments: Any?) {
        guard let args = arguments as? [String: Any],
              let hit = args["hit"] as? [String: Any],
              let model = args["model"] as? [String: Any],
              let path = model["modelPath"] as? String,
              !path.trimmingCharacters(in: .whitespaces).isEmpty
        else { return nil }

        func float(_ map: [String: Any], _ key: String, _ fallback: Float) -> Float {
            guard let value = (map[key] as? NSNumber)?.floatValue, value.isFinite else {
                return fallback
            }
            return value
        }
        func flag(_ key: String) -> Bool { (args[key] as? NSNumber)?.boolValue ?? true }

        hitId = hit["id"] as? String
        position = SIMD3(float(hit, "x", 0), float(hit, "y", 0), float(hit, "z", 0))
        let raw = simd_quatf(
            ix: float(hit, "qx", 0),
            iy: float(hit, "qy", 0),
            iz: float(hit, "qz", 0),
            r: float(hit, "qw", 1)
        )
        rotation = simd_length(raw) > 1e-6 ? simd_normalize(raw) : simd_quatf(ix: 0, iy: 0, iz: 0, r: 1)
        modelPath = path
        let requestedSize = float(model, "scale", 1)
        size = requestedSize > 0 ? requestedSize : 1
        editable = flag("editable")
        draggable = flag("draggable")
        rotatable = flag("rotatable")
        scalable = flag("scalable")
    }
}

/// Clamps a pinched uniform scale to `flutterPlacedScaleRange`.
func flutterClampPlacedScale(_ scale: Float) -> Float {
    guard scale.isFinite else { return 1 }
    return min(max(scale, flutterPlacedScaleRange.lowerBound), flutterPlacedScaleRange.upperBound)
}

File: flutter/sceneview_flutter/android/src/main/kotlin/io/github/sceneview/flutter/ARTapToPlace.kt
package io.github.sceneview.flutter

// ---------------------------------------------------------------------------
// AR tap-to-place (#3780): pure logic, kept free of ARCore and Filament types
// so the JVM unit tests can pin it (ARTapToPlaceTest).
// ---------------------------------------------------------------------------

/**
 * How far a pinch can resize a placed model, relative to the size it was
 * placed at. The Dart doc for `SceneViewController.placeModel` quotes this
 * range, and the iOS bridge clamps to the same one.
 */
internal val PLACED_SCALE_RANGE = 0.25f..4f

/**
 * Maps an ARCore `Plane.Type` name to the plane type string that Dart
 * receives. `onPlaneDetected` and `onPlaneTap` use the same names. [typeName]
 * is `Plane.Type.name`, passed as a string so this stays testable off-device.
 */
internal fun planeTypeName(typeName: String?): String = when (typeName) {
    "HORIZONTAL_UPWARD_FACING" -> "horizontal_upward"
    "HORIZONTAL_DOWNWARD_FACING" -> "horizontal_downward"
    "VERTICAL" -> "vertical"
    else -> "unknown"
}

/**
 * Encodes a plane hit as the `onPlaneTap` payload that the Dart
 * `ARHitResult.fromMap` decodes.
 *
 * @param translation hit position (x, y, z), in metres.
 * @param rotation hit orientation as a quaternion (x, y, z, w). This is the
 * ARCore `Pose.rotationQuaternion` order.
 */
internal fun planeHitMap(
    id: String,
    translation: FloatArray,
    rotation: FloatArray,
    planeType: String,
    distance: Float,
): Map<String, Any> = mapOf(
    "id" to id,
    "x" to translation[0].toDouble(),
    "y" to translation[1].toDouble(),
    "z" to translation[2].toDouble(),
    "qx" to rotation[0].toDouble(),
    "qy" to rotation[1].toDouble(),
    "qz" to rotation[2].toDouble(),
    "qw" to rotation[3].toDouble(),
    "planeType" to planeType,
    "distance" to distance.toDouble(),
)

/** A decoded `placeModel` call. */
internal data class PlaceModelRequest(
    /** Native id of the hit, used to anchor the model to the tapped plane. */
    val hitId: String?,
    val tx: Float,
    val ty: Float,
    val tz: Float,
    val qx: Float,
    val qy: Float,
    val qz: Float,
    val qw: Float,
    val modelPath: String,
    /** Size of the model's largest dimension, in metres (`ModelNode.scale`). */
    val size: Float,
    val editable: Boolean,
    val draggable: Boolean,
    val rotatable: Boolean,
    val scalable: Boolean,
) {
    val canDrag get() = editable && draggable
    val canRotate get() = editable && rotatable
    val canScale get() = editable && scalable
}

/**
 * Decodes the `placeModel` method-channel arguments:
 * `{hit: ARHitResult.toMap(), model: ModelNode.toMap(), editable, draggable,
 * rotatable, scalable}`.
 *
 * Returns `null` when the hit or the model path is missing. Numbers can
 * arrive as `Double`, `Integer` or `Long` depending on how Dart wrote them,
 * so every numeric read goes through [Number].
 */
internal fun parsePlaceModelRequest(args: Map<*, *>?): PlaceModelRequest? {
    val hit = args?.get("hit") as? Map<*, *> ?: return null
    val model = args["model"] as? Map<*, *> ?: return null
    val modelPath = (model["modelPath"] as? String)?.takeIf { it.isNotBlank() } ?: return null
    fun Map<*, *>.float(key: String, fallback: Float) =
        (this[key] as? Number)?.toFloat() ?: fallback
    fun flag(key: String) = args[key] as? Boolean ?: true
    val size = model.float("scale", 1f).takeIf { it > 0f && it.isFinite() } ?: 1f
    return PlaceModelRequest(
        hitId = hit["id"] as? String,
        tx = hit.float("x", 0f),
        ty = hit.float("y", 0f),
        tz = hit.float("z", 0f),
        qx = hit.float("qx", 0f),
        qy = hit.float("qy", 0f),
        qz = hit.float("qz", 0f),
        qw = hit.float("qw", 1f),
        modelPath = modelPath,
        size = size,
        editable = flag("editable"),
        draggable = flag("draggable"),
        rotatable = flag("rotatable"),
        scalable = flag("scalable"),
    )
}

/**
 * Position of a model node, scaled with `scaleToUnits = size`, such that the
 * model's bounding box sits bottom-centred on its parent's origin: centred
 * on X and Z, with its lowest point at Y = 0.
 *
 * The parent is the anchored pivot, so the model stands on the tapped plane
 * and pinch and twist act around the contact point instead of the model's
 * own origin, wherever the source file put it.
 *
 * @param center bounding-box centre in model units (Filament `Box.center`).
 * @param halfExtent bounding-box half extent in model units
 * (Filament `Box.halfExtent`).
 * @param size target size of the largest dimension, in metres. This is the
 * same scale factor `ModelNode.scaleToUnitCube` applies.
 */
internal fun bottomCenterOffset(center: FloatArray, halfExtent: FloatArray, size: Float): FloatArray {
    val maxExtent = maxOf(halfExtent[0], halfExtent[1], halfExtent[2]) * 2f
    val scale = if (maxExtent > 0f) size / maxExtent else 1f
    return floatArrayOf(
        -center[0] * scale,
        (halfExtent[1] - center[1]) * scale,
        -center[2] * scale,
    )
}

/**
 * Keeps the last few plane hits so that a `placeModel` call arriving a few
 * frames after its `onPlaneTap` can still anchor to the plane that was hit.
 * Access is main-thread only. Taps and method calls both arrive there.
 */
internal class RecentHits<T>(private val capacity: Int = 8) {
    private var nextId = 0
    private val hits = object : LinkedHashMap<String, T>(capacity, 0.75f, false) {
        override fun removeEldestEntry(eldest: MutableMap.MutableEntry<String, T>?) =
            size > capacity
    }

    /** Stores [hit] and returns its id. */
    fun put(hit: T): String {
        val id = "hit-${nextId++}"
        hits[id] = hit
        return id
    }

    operator fun get(id: String?): T? = id?.let { hits[it] }

    fun clear() = hits.clear()

    val size get() = hits.size
}
